go/vt/wrangler/reparent.go

```diff
@@ -1043,6 +1026,154 @@ func (wr *Wrangler) emergencyReparentShardLocked(ctx context.Context, ev *events
 	return nil
 }
 
+type replicaStatus struct {
+	replica *topo.TabletInfo
+	status  *replicationdatapb.Status
+}
+
+func (wr *Wrangler) replicaReplicationStatuses(ctx context.Context, keyspace, shard string) ([]*replicaStatus, error) {
+	replicas, err := wr.replicaTabletInfos(ctx, keyspace, shard)
+	if err != nil {
+		return nil, err
+	}
+
+	wg := sync.WaitGroup{}
+	rec := concurrency.AllErrorRecorder{}
+
+	result := make([]*replicaStatus, len(replicas))
+
+	for i, ti := range replicas {
+		wg.Add(1)
+		go func(i int, ti *topo.TabletInfo) {
+			defer wg.Done()
+			status, err := wr.TabletManagerClient().SlaveStatus(ctx, ti.Tablet)
+			if err != nil {
+				rec.RecordError(fmt.Errorf("SlaveStatus(%v) failed: %v", ti.AliasString(), err))
+				return
+			}
+			result[i] = &replicaStatus{
+				replica: ti,
+				status:  status,
+			}
+		}(i, ti)
+	}
+	wg.Wait()
+	return result, rec.Error()
+}
+
+func (wr *Wrangler) replicaTabletInfos(ctx context.Context, keyspace, shard string) ([]*topo.TabletInfo, error) {
+	tabletMap, err := wr.TopoServer().GetTabletMapForShard(ctx, keyspace, shard)
+	if err != nil {
+		return nil, err
+	}
+	allTablets := topotools.CopyMapValues(tabletMap, []*topo.TabletInfo{}).([]*topo.TabletInfo)
+
+	replicas := make([]*topo.TabletInfo, 0, len(allTablets))
+	// Clean up tablets so we only have replicas.
+	for _, ti := range allTablets {
+		if ti.Type == topodatapb.TabletType_REPLICA {
+			replicas = append(replicas, ti)
+		}
+	}
+
+	return replicas, nil
+}
+
+func (wr *Wrangler) CheckTabletIsFarthestAhead(tabletAliasStr string, statusMap map[string]*replicationdatapb.Status) error {
+	tabletStatus, ok := statusMap[tabletAliasStr]
+	if !ok {
+		return fmt.Errorf("couldn't get tablet %v replication position", tabletAliasStr)
+	}
+	candidatePos, err := decodeAndMergePositions(tabletStatus)
+	if err != nil {
+		return fmt.Errorf("cannot decode tablet position %v: %v", tabletStatus.Position, err)
+	}
+	for alias, status := range statusMap {
+		if alias == tabletAliasStr {
+			continue
+		}
+		pos, err := decodeAndMergePositions(status)
+		if err != nil {
+			return fmt.Errorf("cannot decode and merge replica %v executed and retrieved positions %v: %v", alias, status.Position, err)
+		}
+		if !candidatePos.AtLeast(*pos) {
+			return fmt.Errorf("tablet %v is more advanced than master elect tablet %v: %v > %v", alias, tabletAliasStr, status.Position, tabletStatus.Position)
+		}
+	}
+
+	return nil
+}
+
+// FindLatestReplica will search through the status's of the various replicas in the shard,
+// and return the tablet that is the most caught up.
+func (wr *Wrangler) FindLatestReplica(ctx context.Context, keyspace, shard string) (*topodatapb.TabletAlias, error) {
```

Comment:
We have a helper in this file called `maxReplPosSearch`. Can you check if we can dedup with that?

Reply:
Sure, although the name could be better. Since `maxReplPosSearch` is private, can we just rename it and change all calls of it? For a public facing function I would like something that has a clearer name.

Reply:
Yeah it's safe to rename it and otherwise refactor it as needed.

Reply:
I found that `maxReplPosSearch` gets used by `chooseNewMaster` which is a nice name for this kind of a method. I decided to just make this method (on wrangler) public and refactored the `processTablet` method that it utilizes to use the new `decodePosition` helper, so it's using the new union logic.